### PR TITLE
Make downcast macros hygenic (#6400)

### LINKED ISSUE #6400: `downcast_primitive_array` and `downcast_dictionary_array` are not hygienic wrt imports
**Describe the bug**

Using `downcast_primitive_array` and `downcast_dictionary_array` require `arrow_schema` to be in scope of the expanded code.

**To Reproduce**

1. use `arrow::downcast_primitive_array` without `arrow_schema` included as a direct dependency
2. failure to compile due to expanded code using paths from `arrow_schema`

**Expected behavior**

Compile should succeed.

**Additional context**

A workaround is to manually insert something which roughly matches `arrow_schema` into the namespace, e.g.

```rust
    use datafusion::arrow::datatypes as arrow_schema;  // using datafusion facade
    downcast_primitive_array!( ... ); // ^ will now use the above-defined arrow_schema
```

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #6400

# Rationale for this change
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: arrow-array/src/cast.rs [hunk 1]
```diff
@@ -21,6 +21,11 @@ use crate::array::*;
 use crate::types::*;
 use arrow_data::ArrayData;
 
+#[doc(hidden)]
+pub mod __private {
```
Comment (tustvold) [inline arrow-array/src/cast.rs:25]: This is the module name that serde uses for this purpose

Comment (alamb) [inline arrow-array/src/cast.rs:25]: Some context might be nice:

```suggestion
/// Re-export symbols needed for downcast macros
/// 
/// Name follows `serde` convention
#[doc(hidden)]
pub mod __private {
```

**Review by alamb — APPROVED** [review]
Looks good to me -- it might also be worth adding some tests. Perhaps you can follow @crepererum 's model in DataFusion

https://github.com/apache/datafusion/blob/main/datafusion/core/tests/macro_hygiene/mod.rs

**Review by crepererum — APPROVED** [review]

**Comment by crepererum** [comment]
needs a formatting fix

**Comment by alamb** [comment]
I pushed a fix for `fmt` to the branch to get CI passing and will merge after that

**Comment by alamb** [comment]
Thanks again @tustvold and @crepererum

**Pull request merged**